test/apps/network/tcp_poll_err.c

```diff
@@ -0,0 +1,263 @@
+
+	CHECK(shutdown(sk_connect, SHUT_RD));
+	TEST_RES(poll(&pfd, 1, 0),
+		 pfd.revents == (POLLIN | POLLOUT | POLLRDHUP));
+	pfd.fd = sk_connect;
+	TEST_RES(poll(&pfd, 1, 0),
+		 pfd.revents == (POLLIN | POLLOUT | POLLRDHUP));
+}
+END_TEST()
+
+void renew_connect_and_accept()
+{
+	close(sk_connect);
+	close(sk_accept);
+
+	sk_connect = CHECK(socket(PF_INET, SOCK_STREAM, 0));
+	CHECK(connect(sk_connect, (struct sockaddr *)&sk_addr,
+		      sizeof(sk_addr)));
+
+	struct sockaddr addr;
+	socklen_t addrlen = sizeof(addr);
+	sk_accept = CHECK(accept(sk_listen, &addr, &addrlen));
+}
+
+FN_TEST(poll_shutdown_write)
+{
+	renew_connect_and_accept();
+
+	CHECK(write(sk_connect, buf, 1));
+
+	pfd.fd = sk_accept;
+	TEST_RES(poll(&pfd, 1, 0), pfd.revents == (POLLIN | POLLOUT));
+	pfd.fd = sk_connect;
+	TEST_RES(poll(&pfd, 1, 0), pfd.revents == (POLLOUT));
+
+	CHECK(shutdown(sk_accept, SHUT_WR));
+	pfd.fd = sk_accept;
+	TEST_RES(poll(&pfd, 1, 0), pfd.revents == (POLLIN | POLLOUT));
+	pfd.fd = sk_connect;
+	TEST_RES(poll(&pfd, 1, 0),
+		 pfd.revents == (POLLIN | POLLOUT | POLLRDHUP));
+
+	CHECK(write(sk_connect, buf, 1));
+	CHECK(read(sk_accept, buf, 4096));
+
+	TEST_RES(poll(&pfd, 1, 0),
+		 pfd.revents == (POLLIN | POLLOUT | POLLRDHUP));
+
+	CHECK(write(sk_connect, buf, 1));
+	CHECK(read(sk_accept, buf, 4096));
+
+	TEST_RES(poll(&pfd, 1, 0),
+		 pfd.revents == (POLLIN | POLLOUT | POLLRDHUP));
+
+	CHECK(shutdown(sk_connect, SHUT_WR));
+
+	pfd.fd = sk_connect;
+	TEST_RES(poll(&pfd, 1, 0),
+		 pfd.revents == (POLLIN | POLLOUT | POLLRDHUP | POLLHUP));
+	pfd.fd = sk_accept;
+	TEST_RES(poll(&pfd, 1, 0),
+		 pfd.revents == (POLLIN | POLLOUT | POLLRDHUP | POLLHUP));
+}
+END_TEST()
+
+FN_TEST(poll_shutdown_readwrite)
+{
+	renew_connect_and_accept();
+
+	CHECK(write(sk_connect, buf, 1));
+
+	pfd.fd = sk_accept;
+	TEST_RES(poll(&pfd, 1, 0), pfd.revents == (POLLIN | POLLOUT));
+	pfd.fd = sk_connect;
+	TEST_RES(poll(&pfd, 1, 0), pfd.revents == (POLLOUT));
+
+	CHECK(shutdown(sk_accept, SHUT_RDWR));
+	pfd.fd = sk_accept;
+	TEST_RES(poll(&pfd, 1, 0),
+		 pfd.revents == (POLLIN | POLLOUT | POLLRDHUP | POLLHUP));
+	pfd.fd = sk_connect;
+	TEST_RES(poll(&pfd, 1, 0),
+		 pfd.revents == (POLLIN | POLLOUT | POLLRDHUP));
+
+	CHECK(read(sk_connect, buf, 4096));
+	pfd.fd = sk_connect;
+	TEST_RES(poll(&pfd, 1, 0),
+		 pfd.revents == (POLLIN | POLLOUT | POLLRDHUP));
+
+	CHECK(write(sk_connect, buf, 4096));
+
+	// TODO: The following test cannot be passed on Asterinas.
+	// Asterinas does not account for the HUP and ERR results.
+	// It cannot detect when a peer socket is fully closed.
+	// The TCP state machine does not indicate when the peer socket closes its read end.
+	// How does Linux detect this and close the socket after a write syscall?
+	// I've also tested with virtio-net, and its behavior differs from localhost.
+	// The HUP and ERR events are triggered by an RST packet from the peer socket,
+	// but the localhost test does not encounter such an RST packet.
```

Comment:
Could you try to revise the comments to make the following facts clearer?
 1. Linux generates an RST packet in this case, but we currently don't.
 2. RST packets cause `POLERR` in Linux, but we currently have no support for this.

> I've also tested with virtio-net, and its behavior differs from localhost.

Not sure what this sentence whats to mean, but it seems that the comments describe your experiment, which leads the user to reason about what happened in the experiment, and I think that may be unhelpful. We should describe the conclusion directly if we can.